---
title: "unknown-species-id-template"
output: html_notebook
---

28 February 2023


Walkthrough of species id for unknown samples:

Using the rds file output from microhaplot:
1. read in rds files
2. apply read depth filters
3. apply allele balance filter


```{r}
library(tidyverse)
library(readxl)
library(stringr)
library(lubridate)
library(rubias)
library(ggpattern)


source("../R/rockfish-funcs2.R")


#### Call genos from the microhaplot rds files ####

# this next step uses a little data frame that has the names and GTSeq run numbers
# I got that file like this:
# dsb:rds_files dianabaetscher$ pwd
# /Users/dianabaetscher/Desktop/NOAA_grad/git-repos/rockfish-species-id/new_baseline_data/feather_files
# dsb:feather_files dianabaetscher$ ls -l | awk 'BEGIN {print "gtseq_run", "file"} NR >1 && !/posinfo/ {num = $NF; gsub(/gtseq/, "", num); gsub(/[._].*$/, "", num);  print num, $NF}' > ../../data/rds-file-list.txt 

# get the names of the files
fdf <- read.table("../data_AFSC/rds_files/rds-file-list.txt", stringsAsFactors = FALSE, header = TRUE) %>%
  tbl_df()
dir <- "../data_AFSC/rds_files/select_for_analysis/"


# cycle over them, read them and add the gtseq_run column on each.
# at the end, bind them together.
genos_long <- lapply(1:nrow(fdf), function(i) {
  message("Working on ", fdf$file[i])
  call_genos_from_haplotRDS(path = file.path(dir, fdf$file[i])) %>%
    mutate(gtseq_run = fdf$gtseq_run[i]) %>%
    select(gtseq_run, everything())
}) %>%
  bind_rows()

# we go ahead and save it in data/processed, with xz compression
saveRDS(genos_long, file = "../data_AFSC/processed/called_genos.rds", compress = "xz")


#### In the end, let us get a data frame that includes genotypes for all the individuals  ####
# and which explicitly has NAs in places where data are missing, and also 
# has the NMFS_DNA_ID on there
genos_long_explicit_NAs <- genos_long %>%
  select(gtseq_run, id) %>%
  unique() %>%
  unite(col = gid, sep = "_", gtseq_run, id) %>%
  select(gid) %>%
  unlist() %>%
  unname() %>%
  expand.grid(gid = ., locus = unique(genos_long$locus), gene_copy = 1:2, stringsAsFactors = FALSE) %>%
  tbl_df() %>% 
  separate(gid, into = c("gtseq_run", "id"), convert = TRUE) %>%
  left_join(., genos_long) %>%
  arrange(gtseq_run, id, locus, gene_copy)

# and then save that
saveRDS(genos_long_explicit_NAs, file = "../data_AFSC/processed/called_genos_na_explicit.rds", compress = "xz")
  
```


Using those genotypes...
```{r}
genos_long_explicit_NAs %>%
  group_by(gtseq_run, id) %>%
  tally() 

```
381 samples


Look at missing data:
180 gene copies total (90 loci x2)
```{r}
ind_to_toss <- genos_long_explicit_NAs %>%
  group_by(gtseq_run, id) %>%
  filter(is.na(allele)) %>% # missing data
  tally() %>%
  arrange(desc(n)) %>% # remove samples with >20% missing data
  filter(n > 36)

# remove those from the df
genos_ind_filtered <- genos_long_explicit_NAs %>%
  anti_join(., ind_to_toss)
  
```



Load baseline data
```{r baseline-data}
# baseline data - curated, 997 indivs
baseline <- readRDS("../new_baseline_data/processed/sebastes_spp_id_baseline_haplotypes.rds")

# remove the 6 loci that had HWE and other issues
to_remove <- read_csv("../data/loci_to_remove.csv")

baseline90 <- baseline %>%
  anti_join(., to_remove)

# remind myself which species are in the baseline:
baseline90 %>%
  select(collection) %>%
  unique() %>%
  arrange()

tossers <- baseline90 %>%
  select(indiv, gtseq_run, id) %>%
  unique() %>%
  group_by(indiv) %>%
  tally() %>%
  filter(n >1)

baseline90_one_each <- baseline90 %>%
  anti_join(., tossers)

# baseline data - curated, 997 indivs
baseline_spp_info <- baseline90_one_each %>%
  select(sample_type, repunit, collection, indiv, gtseq_run, id, species) %>%
  unique()
baseline_spp_info$gtseq_run <- as.character(baseline_spp_info$gtseq_run)

```



```{r baseline-format}
# slim that down to just the matching field with the unknowns
for_alleidx <- baseline90_one_each %>%
  select(-indiv, -c(1:3, 12:13), -species)
  

for_alleidx$gtseq_run <- as.character(for_alleidx$gtseq_run)
```



```{r}
# merge the two dataframes
merged_df <- bind_rows(for_alleidx, genos_ind_filtered)

# first make integers of the alleles
alle_idxs <- merged_df %>% 
  dplyr::select(gtseq_run, id, locus, gene_copy, allele) %>%
  group_by(locus) %>%
  mutate(alleidx = as.integer(factor(allele, levels = unique(allele)))) %>%
  ungroup() %>%
  arrange(gtseq_run, id, locus, alleidx) # rubias can handle NA's, so no need to change them to 0's

  
# and spread the alleles
two_col <- alle_idxs %>%
  #group_by(indiv, locus) %>%
  unite(loc, locus, gene_copy, sep = ".") %>%
  #ungroup() %>%
  select(-allele) %>%
  pivot_wider(names_from = loc, values_from = alleidx) 


```



add back on info for reference and make two-column format for rubias
```{r}
# baseline
reference <- two_col %>%
  left_join(., baseline_spp_info) %>%
  filter(!is.na(species)) %>%
  select(-gtseq_run, -id, -species) %>%
  select(sample_type, repunit, collection, indiv, everything())

```

```{r}
# mixture
rubias_mix <- two_col %>%
  anti_join(., baseline_spp_info) %>%
  mutate(sample_type = "mixture", collection = "larvae", repunit = NA) %>%
  select(sample_type, repunit, collection, everything()) %>%
  unite(gtseq_run, id, col = "indiv", sep = "_")

```

## Mixture assignment with rubias


```{r run-rubias}
rubias_output <- infer_mixture(reference = reference, mixture = rubias_mix, gen_start_col = 5)

```



```{r}
# take the top output for each sample
top_assign <- rubias_output$indiv_posteriors %>%
  group_by(indiv) %>%
  slice_max(., order_by = PofZ)


top_assign
```

Check on z-scores:

```{r}
top_assign %>%
  ggplot(aes(x = z_score)) +
  geom_histogram()

ggsave("pdf_outputs/larvae_mixed_unknowns_Zscores.pdf")
```

These z-scores suggest that the larvae are not only species not in the baseline, but some of them are very phylogenetically different from the baseline species. 

Some of that assessment depends on what the cherrypicked plates look like:
```{r}
top_assign %>%
  filter(str_detect(indiv, "gtseq2")) %>%
  filter(PofZ > 0.95) %>% # none of the -60 z-scores in this batch
  ggplot(aes(x = z_score)) +
  geom_histogram()
  
```

Read in some metadata to take a closer look at those samples/assignments.

```{r}
POPmix_samplesheet <- read_csv("../data_AFSC/samplesheets/20230110_POPMix_POP_PCOD_BFAL.csv", skip = 19)

meta_spp <- read_xlsx("../data_AFSC/POP_northerns_duskies_to_gtseq.xlsx")
meta_spp$ABLG <- as.character(meta_spp$ABLG)

# join those to get the species
meta_for_gtseqMix <- POPmix_samplesheet %>%
  filter(Sample_Project == "POPMix") %>%
       left_join(., meta_spp, by = c("Sample_ID" = "ABLG")) %>%
  select(Sample_ID, SpeciesName, AlternateID_s_, Sample_Plate, Sample_Well) %>%
  mutate(id = row_number()) %>%
  mutate(s = "s") %>%
  unite(col = "s_num", s, id, sep = "")
```


```{r}
# POP and duskies mixed plate
spp_id_w_knowns <- top_assign %>%
  filter(str_detect(indiv, "gtseq2")) %>%
  separate(indiv, into = c("gtseq_run", "id")) %>%
  left_join(., meta_for_gtseqMix, by = c("id" = "s_num")) %>%
  select(collection, SpeciesName, z_score, everything()) %>%
  rename(rubias_assignment = collection)

spp_id_w_knowns %>%
  ggplot(aes(x = z_score, fill = SpeciesName)) +
  geom_histogram(alpha = 0.3) +
  theme_bw()


spp_id_w_knowns %>%
  ggplot(aes(x = z_score, fill = SpeciesName)) +
  geom_histogram(alpha = 0.5) +
  facet_grid(rows = vars(SpeciesName)) +
  theme_bw() +
  theme(
    strip.text.y = element_text(size = 8)) +
  labs(title = "POP and dusky mixed plate - species ID")

ggsave("pdf_outputs/known_sebastes_POP_dusky_northern_sppID.pdf", height = 5, width = 5)
```

```{r}
spp_id_w_knowns %>%
  select(-missing_loci) %>%
  write_csv("csv_outputs/known_northern_POP_dusky_rubias_id.csv")

spp_id_w_knowns %>%
  filter(SpeciesName == "Sebastes polyspinis")
```


That actually makes me feel considerably better about the assignment. The z-scores for Northerns look normal (and Northerns are panmictic) whereas the POP has a wider distribution (because of pop structure, likely) and the duskies are the first set of z-score outliers.

In fact, we could use the z-scores to some extent, to give us a hint about the larval samples.

```{r}
goa <- top_assign %>%
  filter(!str_detect(indiv, "gtseq2")) 

# combined plot
ggplot() +
  geom_histogram(data = goa, aes(x = z_score), alpha = 0.5) +
  geom_histogram(data = spp_id_w_knowns, aes(x = z_score, fill = SpeciesName), alpha = 0.25) +
  theme_bw() +
  labs(title = "GOA unknown larvae - Sebastes species ID")

ggsave("pdf_outputs/GOA_unknowns_w_unknownSpp2.pdf")
```
What was the species assignment for the -60 bunch?

```{r}
goa %>%
  filter(z_score < -20)

```







```{r}
# those samples with normal parameters
normal_Zs <- top_assign %>%
  filter(z_score > -3 & z_score < 3) 

normal_Zs %>%
  ggplot(aes(x = z_score)) +
  geom_histogram()
  

```


How many larvae from which species in the "normal" range for baseline species?

```{r}
normal_Zs %>%
  group_by(collection) %>%
  tally()

```
POP, northerns, quillback, redbanded, and tiger rockfishes. All reasonable.













## Check self-assignment of baseline

Sanity-check

Look at self-assignment:

```{r}
self_test <- self_assign(reference = reference, gen_start_col = 5)
```

```{r}
self_test %>%
  group_by(indiv) %>%
  slice_max(., order_by = scaled_likelihood) %>%
  ggplot(aes(x = z_score)) +
  geom_histogram()
  
ggsave("pdf_outputs/baseline_zscores.pdf")
```


Self-assignment looks fine. 

I'm beginning to think that what we're seeing in the z-scores is (1) the population structure in POP, (2) duskies missing from the baseline, and (3) some weird species - maybe not sebastes?? in the larval collections.








##### DO NOT MODIFY BELOW THIS LINE #####



## New primer pool


Take a look at the distribution of reads across loci:

```{r}
loc_depth <- genos_long_explicit_NAs %>%
  group_by(gtseq_run, id, locus) %>%
  filter(!is.na(depth)) %>% # remove the NAs, which interfere with sum
  ungroup() %>%
  group_by(locus) %>%
  summarise(mean_depth = mean(depth)) %>%
  arrange(desc(mean_depth)) 

loc_depth %>%
  ggplot(aes(x = reorder(locus, mean_depth), y = mean_depth)) +
  geom_bar(stat = "identity") +
  theme(
    axis.text.x = element_text(angle = 90, hjust = 0.95)
  )

```


Try changing some primer concentrations:

C1* V1 = C2 *V2

```{r}
primer_conc <- loc_depth %>%
  arrange(desc(mean_depth)) %>%
  mutate(primer_multiplier = ifelse(mean_depth < 100, 2, 1)) %>%
  mutate(primer_multiplier = ifelse(mean_depth > 500, 0.5, primer_multiplier)) %>%
  mutate(primer_conc = 0.25) %>%
  mutate(new_primer_conc = primer_multiplier*primer_conc) %>%
  mutate(C1 = 200) %>%
  mutate(V2 = 1500) %>%
  mutate(V1 = round(V2*new_primer_conc/C1)) %>%
  rename(vol_to_add = V1, total_vol = V2)
  # summarise(sum(vol_to_add))
 

```
202 ul of primers

```{r}
1500-202 
```
1298 ul of Tris or water to add.


Read in primer order for plate map:
```{r}
primerOrder <- read_xlsx("../../../GTseq/sebastesSppID/PrimerOrder.xlsx")

primer_conc$locus <- gsub("tag_id", "Sat", primer_conc$locus)

plate_primers <- primer_conc %>%
  filter(str_detect(locus, "Plate_")) %>%
  separate(locus, into = c("plate", "n", "well", "spp", "label", "drop")) %>%
  unite(col = "locus", spp, label) %>%
  select(-c(1:3, 5))



primerOrder$`Sequence Name` <- gsub("_F", "", primerOrder$`Sequence Name`)
order_short <- primerOrder %>%
  select(`Well Position`, `Sequence Name`)


primer_conc %>%
  filter(!str_detect(locus, "Plate_")) %>%
  bind_rows(., plate_primers) %>%
  left_join(., order_short, by = c("locus" = "Sequence Name")) %>%
  unique() %>%
   write_csv("csv_outputs/sebastes_spp_id_primerPool_20230228.csv")
  
```

